### PR TITLE
chore(storage): add server mode [benchmarks]

### PR DESCRIPTION
**Pull request by BrennaEpp**

### CODE REVIEW

**Comment by quartzmo** [comment]
Typo in the title of this PR: "shore"

File: storage/internal/benchmarks/main.go [hunk 1]
```diff
@@ -21,12 +21,15 @@ import (
...
 	"log"
-	"math/rand"
+	"net/http"
 	"os"
 	"strconv"
 	"strings"
 	"time"
 
+	epb "github.com/cloudprober/cloudprober/probes/external/proto"
+	"github.com/cloudprober/cloudprober/probes/external/serverutils"
```
Comment (tritone) [inline storage/internal/benchmarks/main.go:31]: Fix export ordering

File: storage/internal/benchmarks/main.go [hunk 2]
```diff
@@ -225,7 +236,12 @@ func parseFlags() {
 func main() {
 	log.SetOutput(os.Stderr)
 	parseFlags()
-	rand.Seed(time.Now().UnixNano())
+
+	go func() {
+		if err := http.ListenAndServe("localhost:8080", nil); err != nil {
```
Comment (tritone) [inline storage/internal/benchmarks/main.go:241]: What does this do as a stand-alone? add a comment?

File: storage/internal/benchmarks/main.go [hunk 3]
```diff
@@ -269,96 +286,33 @@ func main() {
...
-				// to indicate that we did see an error, even though we continue.
-				// Cleanup may be expected to fail if there is an issue with the run.
-				log.Printf("run cleanup failed: %v", err)
-				exitWithErrorCode = true
-			}
-			return nil
-		})
+	if serverMode {
+		// Server mode blocks forever servicing probe requests
+		runInServerMode(ctx, opts)
```
Comment (tritone) [inline storage/internal/benchmarks/main.go:304]: Should this have an exit()? If runInServerMode does return then we would probably not want to continue right?

File: storage/internal/benchmarks/main.go [hunk 4]
```diff
@@ -494,6 +448,139 @@ func startRecordingResults(w io.Writer, g *errgroup.Group, oType outputType) {
...
+
+// runInServerMode blocks forever servicing probe requests.
+// Number of samples requested is ignored in this mode.
+// Timeouts must be properly set at the probe level to ensure requests aren't
+// being serviced concurrently
+func runInServerMode(ctx context.Context, opts *benchmarkOptions) {
+	results = make(chan benchmarkResult, 4)
+
+	serverutils.Serve(func(request *epb.ProbeRequest, reply *epb.ProbeReply) {
+		timeLimit := time.Millisecond * time.Duration(request.GetTimeLimit())
```
Comment (tritone) [inline storage/internal/benchmarks/main.go:542]: Maybe note that this is per-sample?

**Review by tritone — COMMENTED** [review]
couple questions, looks good overall!

File: storage/internal/benchmarks/main.go (hunk 1 above)
Comment (BrennaEpp) [reply storage/internal/benchmarks/main.go:31]: What do you mean? Should I ungroup these?

File: storage/internal/benchmarks/main.go (hunk 3 above)
Comment (BrennaEpp) [reply storage/internal/benchmarks/main.go:294]: That is correct; although it should never return, I've added a check here to make sure we do error out if it does.

File: storage/internal/benchmarks/main.go (hunk 4 above)
Comment (BrennaEpp) [reply storage/internal/benchmarks/main.go:538]: Added a comment; let me know if it makes sense or is still missing context.

File: storage/internal/benchmarks/main.go (hunk 2 above)
Comment (BrennaEpp) [reply storage/internal/benchmarks/main.go]: Whoops, this was for obtaining pprof samples. I removed it; I'll open a PR separately to see if it makes more sense to wrap specific parts of the code with cpu profiling and take periodic memory samples, or else re-add this back in for ad-hoc profile collecting.

File: storage/internal/benchmarks/main.go (hunk 1 above)
Comment (tritone) [reply storage/internal/benchmarks/main.go:30]: Oh I guess it's just weird that these are grouped separately; they can go with the other non-core imports below.

**Review by tritone — APPROVED** [review]

**Pull request merged**